### PR TITLE
Add a ContainerRuntime adapter; example of load runtime with extra blob, and testing for sample use

### PR DESCRIPTION
**Pull request by chensixx**
## Description
Adding a containerRuntime adapter, `mixinSummaryHandler` which returns containerRuntime with overwritten method. in this PR, `summarize` is overwritten to add a blob to summary.
The new test in `summaries.spec.ts` is a sample to show how to use the adapter. We can load a containerRuntime that has extraBlob in summary. the test also verifies blob content is right.

Task 3117: https://dev.azure.com/fluidframework/internal/_workitems/edit/3117/

### CODE REVIEW

**Comment by chensixx** [comment]
I'm not sure if unit test is enough since I have to call summarize to verify we have the right blob. So I also added the e2e test

**Comment by agarwal-navin** [comment]
@chensixx Can you please add a description?

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -3458,3 +3462,61 @@ const waitForSeq = async (
...
+		for (const name of path) {
+			if (baseSnapshot) {
+				baseSnapshot = baseSnapshot?.trees[name];
+			}
+		}
+
+		if (baseSnapshot && blobName) {
+			// We got the blob!
+			assert(storage !== undefined, "attached state should have storage");
+			extraBlob = await readAndParse<string>(storage, baseSnapshot?.blobs[blobName]);
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3516]: should we just return here when we find it? And return 'undefined` after the topmost 'if'? We'd get rid of the `extraBlob` variable.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -3458,3 +3462,61 @@ const waitForSeq = async (
...
+			console.log(extraBlob?.length ?? 0);
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3517]: Not sure if this is intentional. If you need to log this event, use the logger and be mindful of the volume of logs (consider sampling it) but the console goes nowhere.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -3458,3 +3462,61 @@ const waitForSeq = async (
...
+	const context = params.context;
+	const pendingRuntimeState = context.pendingLocalState as IPendingRuntimeState | undefined;
+	let baseSnapshot = context.baseSnapshot;
+	const storage = context.storage;
+
+	// Load custom application blob
+	// This state is only available for real snapshots, we do not write that extra state when serializing local (uncommited) changes.
+	let extraBlob: string | undefined;
+	if (pendingRuntimeState === undefined) {
+		const blobName = path.pop();
```
Comment (andre4i) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3506]: `path.pop()` mutates the path array, modifying the array provided by the caller of the function. Do really want to do that?

Comment (chensixx) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3506]: i see, i will use slice instead

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 4]
```diff
@@ -3458,3 +3462,66 @@ const waitForSeq = async (
 			deltaManager.on("op", handleOp);
 		}
 	});
+
+/**
+ * Mixin class that adds await for DataObject to finish initialization before we proceed to summary.
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3467]: The comment is incorrect.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 5]
```diff
@@ -422,6 +423,17 @@ export interface IRootSummaryTreeWithStats extends ISummaryTreeWithStats {
 	gcStats?: IGCStats;
 }
 
+export interface IContainerRuntimeParams {
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:426]: IContainerRuntimeCreateParams is a more apt name. Also, please add a comment that indicates what it is.

File: packages/runtime/runtime-utils/src/summaryUtils.ts [hunk 6]
```diff
@@ -116,6 +117,34 @@ export function addBlobToSummary(
 	summary.stats.totalBlobSize += getBlobSize(content);
 }
 
+export function addBlobToSummaryViaPath(
```
Comment (agarwal-navin) [inline packages/runtime/runtime-utils/src/summaryUtils.ts:120]: This function needs comments with example. It is not at all clear how it adds the given blob to the summary.

Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summaryUtils.ts:120]: It also needs unit tests. I can see this easily breaking.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 7]
```diff
@@ -3458,3 +3462,66 @@ const waitForSeq = async (
...
+					content.content,
+				);
+			}
+			return result;
+		}
+	} as typeof ContainerRuntime;
+
+// Helper API to load extra blobs in snapshot
+// This API only works for blobs that are saved in storage, i.e. this API would only work for load from storage scenarios
+export async function loadRuntimeBlob(
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3498]: What is this for? Looks like its only used in a test and should move to that test. It does not look like something we would like to be part of the public API.

File: packages/test/test-end-to-end-tests/src/test/summaries.spec.ts [hunk 8]
```diff
@@ -192,6 +196,65 @@ describeNoCompat("Summaries", (getTestObjectProvider) => {
 		assert(ackNackResult.success, "summary op should be acked");
 	});
 
+	it("sample use for customized containerRuntimeWithBlob", async () => {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summaries.spec.ts:199]: This should be a unit test in container runtime. There is nothing end-to-end about it :-).

**Comment by agarwal-navin** [comment]
@chensixx Why is this mixin added? What is the use case for adding a blob to the container runtime summary? It seems very hacky and dangerous.

File: packages/test/test-end-to-end-tests/src/test/summaries.spec.ts [hunk 9]
```diff
@@ -192,6 +196,65 @@ describeNoCompat("Summaries", (getTestObjectProvider) => {
...
+		const blobContents = "sample payload";
+
+		// sample API to load a containerRuntime which has extra blob in the summary
+		async function containerRuntimeWithBlob(params: IContainerRuntimeParams) {
+			// Blob path in container snapshots
+			const path = ["loop", "guestComponents"];
+
+			const extraBlob = await loadRuntimeBlob(params, path);
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/summaries.spec.ts:207]: Why is this step needed during container runtime creation? extraBlob is not even used.

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 7 above)
Comment (chensixx) [reply packages/runtime/container-runtime/src/containerRuntime.ts:3504]: i think this is for loading a blob immediately after the containerRuntime is booted. for loop to faster load a blob they want

File: packages/runtime/container-runtime/src/test/containerRuntime.spec.ts [hunk 10]
```diff
@@ -87,6 +89,51 @@ describe("Runtime", () => {
...
+				const result = await runtimeWithBlob.summarize({
+					runGC: false,
+					fullTree: false,
+					trackState: false,
+					summaryLogger: new MockLogger(),
+				});
+		
+				const blobTree = (result.summary.tree[".application"] as ISummaryTree).tree;
+				const blob = (blobTree.loop as ISummaryTree).tree.guestComponents as ISummaryBlob;
+				assert.strictEqual(blob.content, blobContents, "blob has the wrong content");
```
Comment (chensixx) [inline packages/runtime/container-runtime/src/test/containerRuntime.spec.ts:133]: also need to test loadRuntimeBlob here

**Pull request closed without merging**